### DISCUSSION TITLE
How to get the armor of private key after decrypting it without losing passphrase?

### DISCUSSION THREAD

**Discussion by Joakal** [openpgpjs/openpgpjs#1501] [General]
As title says.

Context: I wondered why I kept getting "Key packet is already encrypted" error. I've done it like this:

```
const privateKeyArmored = `-----BEGIN PGP PRIVATE KEY BLOCK-----
...
-----END PGP PRIVATE KEY BLOCK-----`; // encrypted private key
const passphrase = `yourPassphrase`; // what the private key is encrypted with

const privateKey = await openpgp.decryptKey({
    privateKey: await openpgp.readPrivateKey({ armoredKey: privateKeyArmored }),
    passphrase
});

// It passes!

// Lets save privateKey and publicKey for user

const keyStore = {privateKey: privateKey.armor(), publicKey: privateKey.toPublic().armor()}

// Application offers to user if they want to download private/public key?
console.log(keyStore.privateKey); // different to privateKeyArmored because passphrase was removed
```

Thankfully, OpenPGP package said "Key packet is already encrypted" otherwise, the application would stupidly store passphrase-less private keys.

**Comment by larabr** [comment]
Hi @Joakal , I'm not sure I understand your question.
What operation gives you "Key packet is already encrypted"?

Calling `.armor()` on a decrypted key does not automatically re-encrypt the key. After decrypting a key, you need to manually re-encrypt it using `openpgp.encryptKey`, and then armor the returned value, e.g.:
```js
const encryptedKey = await openpgp.encryptKey({ privateKey, passphrase });
console.log(encryptedKey.armor()) // this is the encrypted armored key
```
Note that `privateKey` will still be decrypted though (`encryptKey` does not change the key given as input). If you want to double check whether a private key is decrypted, you can call `.isDecrypted()` on it.

**Reply by Joakal** [reply]
I was under the impression that decrypting the key would also have the original when I do armor. Not a problem, I'll keep the original armor separately. 

Probably best to do a warning(?) that the armor() on a private key is returning the armor without the passphrase, as it may trip up some developers?

**Reply by twiss** [reply]
> the armor() on a private key is returning the armor without the passphrase

It doesn't, it returns it however the key object is. If the key object is decrypted, the armor is too; if the key object is encrypted, so is the armored key. So you also don't necessarily have to keep the original armor; if you don't want, you can just keep the encrypted key object:

```js
const privateKey = await openpgp.readPrivateKey({ armoredKey: privateKeyArmored });
privateKey.armor(); // armored encrypted key

const decryptedKey = await openpgp.decryptKey({ privateKey, passphrase });
decryptedKey.armor(); // armored decrypted key
```
